splash/browser_tab.py

```diff
@@ -503,6 +511,64 @@ def runjs(self, js_source, handle_errors=True):
         js_source = "%s;undefined" % js_source
         self.evaljs(js_source, handle_errors=handle_errors)
 
+    def wait_for_resume(self, js_source, callback, errback, timeout):
+        """
+        Run some Javascript asynchronously.
+
+        The JavaScript must contain a method called `main()` that accepts
+        one argument. The first argument will be an object with `resume()`
+        and `error()` methods. The code _must_ call one of these functions
+        before the timeout or else it will be canceled.
+
+        Note: this cleans up the JavaScript global variable that it creates,
+        but QT seems to notice when a JS GV is deleted and it destroys the
+        underlying C++ object. Therefore, we can only delete the JS GV _after_
+        the user's code has called us back. This should change in QT5, since
+        it will then be possible to specify a different object ownership
+        policy when calling addToJavaScriptWindowObject().
+        """
+
+        frame = self.web_page.mainFrame()
+        script_text = js_source.replace('\n', ' \\\n').replace('\"', '\\"')
+        callback_proxy = OneShotCallbackProxy(self, callback, errback, timeout)
+        frame.addToJavaScriptWindowObject(callback_proxy.name, callback_proxy)
+
+        wrapped = """
+        (function () {
+            try {
+                eval("%(script_text)s");
+            } catch (err) {
+                var main = function (splash) {
+                    splash.error(err);
```

Comment:
I did some research on it this morning. `QObject` does not work in the `pyqtSlot` decorator. QT will say:

> Error: incompatible type of argument(s) in call to resume(); candidates were
> resume(QObject*)

I get a similar message if I try Python's `object`.

> Error: incompatible type of argument(s) in call to resume(); candidates were
> resume(PyQt_PyObject)

I finally found that `QVariant` works: it allows QT to call the resume() method. It sometimes produces QT types (e.g. QString) that are not useful in Lua (b/c they are not converted to Lua types automatically) and break JSON serialization if you try to include one in your response. We can work around by adding logic to recursively convert QT types to built-in Python types, and that supports nil, boolean, number, and string types correctly.

The only hitch is that a JavaScript `Array` cannot be passed to a slot that expects `QVariant`. I get this error:

> Error: incompatible type of argument(s) in call to resume(); candidates were
> resume(QVariant)

I can add an overload of `resume()` that declares a `QList` argument:

    @pyqtSlot('QList<QVariant>', name='resume')
    def does_this_work(self):
        print('yes it works')

This works for int, bool, and array, but it breaks strings! If I pass a string:

> Error: ambiguous call of overloaded function resume(); candidates were
> resume(QVariant)
> resume(QList<QVariant>)

Arghhh... QT is frustrating.

I think the only way to make this work is to explicitly overload `resume()` for each C++ type that we support, e.g. `float`, `bool`, `void`, etc. `QList`, etc. This is a more verbose and a less flexible because we can't rely on QVariant's type conversion magic, but at least it works. Here's an example:

    @pyqtSlot
    @pyqtSlot('bool')
    @pyqtSlot('double')
    @pyqtSlot('QByteArray')
    ... etc ...
    def resume(self, value):
        ... convert value to built-in python type and then do resume stuff ...

The alternative is to modify the JS wrapper in `browser_tab.py:wait_for_resume()` to serialize arguments in JSON format and then decode that JSON in our Python handler. The code would be shorter and more general than using the overloads described above. I think it is hacky, but I found a post on [StackOverflow](http://stackoverflow.com/a/23550868/122763) that prefers using JSON. What do you think?

I implemented this today using overloads instead of JSON, but let me know if you prefer the JSON approach. I don't think it would be too hard to try it out, and the test coverage is good enough that I would quickly find out if it works well.